[ticket/16708] Update the UTF-8 check for multiple encoding settings

EA117 opened the pull request:
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [ ] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

https://tracker.phpbb.com/browse/PHPBB3-16708

EA117 committed:
[ticket/16708] Update the UTF-8 check for multiple encoding settings

Update the mbstring.http_input and mbstring.http_output configuration check, from being the deprecated PHP 5.6.x "PASS, or blank" to now be "UTF-8, or blank".  PHP documentation specifies this configuration should be left blank, in order to inherit default_charset, which phpBB requires to be configured as UTF-8.

For a phpBB customer where mbstring.http_input and mbstring.http_output is set to UTF-8 rather than being left blank, and is a shared hosting configuration that the customer isn't permitted to change, the phpBB ACP warning that "your http_input/http_output configuration is invalid" isn't helpful.

Their setting is UTF-8, which is the encoding http_input/http_output would be using if it had been left blank and allowed to inherit default_charset.  Our "warning" isn't advising the user of a misconfiguration which is causing a phpBB functionality problem.  phpBB is getting exactly the UTF-8 encoding it was intending to get; same as when it was checking whether the setting was left blank, too.

Also update the default_charset test to permit "UTF-8, or blank".  This had been correctly changed to effectively test "UTF-8, or NULL" during PHPBB3-16698, based on code inspection of /vendor/bantu/ini-get-wrapper/src/IniGetWrapper.php which confirmed IniGetWrapper::getString() will return NULL for any setting which does not exist.

However, testing on some PHP platforms shows PHP itself returning an empty string for a parameter which isn't defined in the PHP.INI, and so we receive an empty string instead of NULL.  So the test was updated so that we effectively check for "UTF-8, or blank, or NULL", all of which result in UTF-8 being used.

PHPBB3-16708

rxu commented on phpBB/includes/acp/acp_main.php:
```diff
@@ -702,9 +702,9 @@ function main($id, $mode)
-				'S_MBSTRING_HTTP_INPUT_FAIL'			=> !empty($http_input),
-				'S_MBSTRING_HTTP_OUTPUT_FAIL'			=> !empty($http_output),
-				'S_DEFAULT_CHARSET_FAIL'				=> $default_charset !== null && strtolower($default_charset) !== 'utf-8',
+				'S_MBSTRING_HTTP_INPUT_FAIL'			=> $http_input !== null && !in_array(strtolower($http_input), array('utf-8', '')),
+				'S_MBSTRING_HTTP_OUTPUT_FAIL'			=> $http_output !== null && !in_array(strtolower($http_output), array('utf-8', '')),
+				'S_DEFAULT_CHARSET_FAIL'				=> $default_charset !== null && !in_array(strtolower($default_charset), array('utf-8', '')),
```
Short array syntax, please :)

CHItA replied:
And encapsulate it into a function/method/lambda/class/death star.

EA117 replied:
Thanks, on both points, considered to be addressed with latest commit.

EA117 committed:
[ticket/16708] Update the UTF-8 check for multiple encoding settings

Fixed short array syntax, moved testing logic into member helper function.

PHPBB3-16708

3D-I commented:
I am not a fan of this changes.

Neustradamus commented:
@marc1706: What do you think about this PR?
